File: revisiting_ctci/OneAway_1_5.go
/*

given two strings, write a method to determine if they are one (or less)
change (insertion, deletion, replacement) away from one another

*/


package main

import (
	"fmt"
)


// take two strings and compare them to one another to see if they are one
// insertion, deletion or replacement away from one another
func OneAway(s1, s2 string) bool {
	diff := 0
	long := ""
	short := ""
	
	// check for replacement first, only valid if same length
	if len(s1) == len(s2){
		for i:=0 ; i < len(s1) ; i++ {
			if s1[i] != s2[i]{
				diff++
			}
			if diff > 1 {
				return false
			}
		}

	// if they aren't the same size only one indel is allowed
	// check to see if they are one different in length, determine the larger
	} else if (len(s1) - len(s2)) == 1 {
		long = s1
		short = s2
	} else if (len(s1) - len(s2)) == -1 {
		long = s2
		short = s1
	} else {
		return false
	}

	offset := 0
	for i := 0  ; i < len(short); i++ {
		li := i + offset
		if short[i] != long[li] {
			offset++
			diff++
		}
		if diff > 1 {
			return false
		}
	}

	return true
}




func main() {

	t1 := "TATTA"
	t2 := "TACTTA"
	fmt.Printf("%v\t%v\t%v\n", t1, t2, OneAway(t1,t2))

	t3 := "CGCGCG"
	t4 := "CGCTCG"
	fmt.Printf("%v\t%v\t%v\n", t3, t4, OneAway(t3,t4))


	t5 := "ATGCATGC"
	t6 := "ATGCTTATGC"
	fmt.Printf("%v\t%v\t%v\n", t5, t6, OneAway(t5,t6))


	t7 := "GGGGTAC"
	t8 := "GGCCTAC"
	fmt.Printf("%v\t%v\t%v\n", t7, t8, OneAway(t7,t8))


}

File: revisiting_ctci/OneAway_1_5.py
import unittest

"""
given two strings, write a method to determine if they are one (or less)
change (insertion, deletion, replacement) away from one another
"""

def OneAway(s1, s2):
	""" take two input strings, return true 
		if they are one change different from one another,
		false if > 1 difference """
	diff = 0

	if len(s1) == len(s2):
		for pos, _ in enumerate(s1):
			if s1[pos] != s2[pos]:
				diff += 1
			if diff > 1:
				return False

	elif (len(s1) - len(s2) == 1) or (len(s1) - len(s2) == -1):
		if len(s1) >  len(s2):
			longer = s1
			shorter = s2
		elif len(s1) < len(s2):
			longer = s2
			shorter = s1
		offset = 0
		for pos, _ in enumerate(shorter):
			if shorter[pos] != longer[pos+offset]:
				diff += 1
				offset += 1
			if diff > 1 :
				return False

	else:
		return False

	return True

class TestOne(unittest.TestCase):
	def test_OneAway(self):
		self.assertTrue(OneAway('timmy', 'jimmy'))
		self.assertTrue(OneAway('dale', 'daleo'))

		self.assertFalse(OneAway('chinook', 'think'))
		self.assertFalse(OneAway('doug', 'dave'))


if __name__ == '__main__':


	unittest.main()